### PR TITLE
[FJTN-52] stats monitor

### PR DESCRIPTION
**Pull request by kyrov**
- add stats api
- add ability to comsume reserve data from rabbitMQ
- still not done the download function
- I suffer and need sleep
- see u at around 14:00

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1 +1,22 @@
...
+To run the mock data, run the following command:
+
+```bash
+cd test_rabbitmq
+python producer.py
+```
+
+## get stats api
+
+Get stats api is on the /api/reports/{space_name}
```
Comment (phumipatc) [inline README.md:20]: `space_name` or `space_id` ?

File: router/report.go [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+import (
+	"github.com/FakJeongTeeNhoi/report-system/controller"
+	"github.com/gin-gonic/gin"
+)
+
+func ReportRouterGroup(server *gin.RouterGroup) {
+	report := server.Group("/report")
+	{
+		report.GET("/:space", controller.GetSpaceStatistic)
```
Comment (phumipatc) [inline router/report.go]: specify `:space` ว่ามันคือ `spaceId` หรือ `spaceName`

File: main.go [hunk 3]
```diff
@@ -27,8 +32,10 @@ func main() {
 	server.Use(cors.New(corsConfig))
 
 	api := server.Group("/api")
+	go controller.StartConsumeDataFromQueue("Receiver", []string{"topic"})
```
Comment (phumipatc) [inline main.go]: Should routing key be `reservation.*` for scalability? Example events are `reservation.created` and `reservation.updated`.

File: test_rabbitmq/producer.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+# Convert report to JSON
+report_json = json.dumps(report)
+
+connection = pika.BlockingConnection(
+    pika.ConnectionParameters(host='localhost',port=5672))
+channel = connection.channel()
+
+channel.exchange_declare(exchange='Receiver', exchange_type='topic', durable=True)
+
+routing_key = "topic"
```
Comment (phumipatc) [inline test_rabbitmq/producer.py]: If you change the defined routing key, don't forget to also change this line.

File: go.mod [hunk 5]
```diff
@@ -1,11 +1,14 @@
-module github.com/FakJeongTeeNhoi/template
+module github.com/FakJeongTeeNhoi/report-system
```
Comment (phumipatc) [inline go.mod]: Wrong module name. It should be `github.com/FakJeongTeeNhoi/report-service` according to the repo name

Comment (phumipatc) [reply go.mod]: If you change this line don't forget to change the import section in other file

File: controller/report.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
+package controller
+
+import (
+	"github.com/FakJeongTeeNhoi/report-system/model"
+	"github.com/FakJeongTeeNhoi/report-system/model/response"
+	"github.com/gin-gonic/gin"
+)
+
+func GetSpaceStatistic(c *gin.Context) {
+	space := c.Param("space")
```
Comment (phumipatc) [inline controller/report.go]: specify `:space` ว่ามันคือ `spaceId` หรือ `spaceName`

File: controller/consumer.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+func failOnError(err error, msg string) {
+	if err != nil {
+		log.Panicf("%s: %s", msg, err)
+	}
+}
+
+func StartConsumeDataFromQueue(Exchange string, key []string) {
+
+	conn, err := amqp.Dial("amqp://guest:guest@rabbitmq:5672/")
```
Comment (phumipatc) [inline controller/consumer.go]: The connection string should be in `.env` for security purpose.

File: model/report.go [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+type Participant struct {
+	Type    string `json:"type"`
+	Faculty string `json:"faculty"`
+}
+
+type Report struct {
+	Id            uuid.UUID     `json:"id"`
+	ReservationId uuid.UUID     `json:"reservation_id"`
+	RoomId        uuid.UUID     `json:"room_id"`
+	Space         string        `json:"space"`
```
Comment (phumipatc) [inline model/report.go]: `spaceId` or `spaceName`?

File: model/report.go [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+
+type Reserve struct {
+	Id            uuid.UUID     `json:"id"`
+	Participant   []Participant `json:"participant"`
+	Status        string        `json:"status"`
+	Approver      string        `json:"approver"`
+	StartDatetime time.Time     `json:"start_datetime"`
+	EndDatetime   time.Time     `json:"end_datetime"`
+	RoomId        uuid.UUID     `json:"room_id"`
+	Space         string        `json:"space"`
```
Comment (phumipatc) [inline model/report.go]: `spaceId` or `spaceName`?

**Review by phumipatc — CHANGES_REQUESTED** [review]
reviewed.

**Comment by phumipatc** [comment]
JIRA task also should be `FJTN-52`.

File: README.md (hunk 1 above)
Comment (kyrov) [reply README.md:20]: space_name

File: router/report.go (hunk 2 above)
Comment (kyrov) [reply router/report.go]: space_name

File: controller/report.go (hunk 6 above)
Comment (kyrov) [reply controller/report.go]: spaceName

File: main.go (hunk 3 above)
Comment (kyrov) [reply main.go]: done

File: test_rabbitmq/producer.py (hunk 4 above)
Comment (kyrov) [reply test_rabbitmq/producer.py]: done

File: go.mod (hunk 5 above)
Comment (kyrov) [reply go.mod]: done

File: controller/consumer.go (hunk 7 above)
Comment (kyrov) [reply controller/consumer.go]: done

**Comment by kyrov** [comment]
all solve and running
pls check again

**Comment by kyrov** [comment]
- get rid of UUID (use string instead)
- querry using spaceID instead of spaceName
- test and work

**Review by phumipatc — APPROVED** [review]

**Pull request merged**